packages/cache/src/AzureBlobCacheStorage.ts

```diff
@@ -29,41 +31,67 @@ function createBlobClient(
 }
 
 export class AzureBlobCacheStorage extends CacheStorage {
-  constructor(private options: AzureBlobCacheStorageOptions) {
+  constructor(
+    private options: AzureBlobCacheStorageOptions,
+    private internalCacheFolder: string
+  ) {
     super();
   }
 
-  protected _fetch(hash: string, destinationFolder: string): Promise<boolean> {
-    const blobClient = createBlobClient(
-      this.options.connectionString,
-      this.options.container,
+  protected async _fetch(hash: string): Promise<string | undefined> {
+    const temporaryBlobOutputFolder = path.join(
+      this.internalCacheFolder,
+      "azure-blob",
       hash
     );
 
-    return blobClient
-      .download(0)
-      .then(response => {
-        const parentFolderWhereToExtractFolder = path.join(
-          destinationFolder,
-          ".."
+    if (!fs.existsSync(temporaryBlobOutputFolder)) {
+      try {
+        const blobClient = createBlobClient(
+          this.options.connectionString,
+          this.options.container,
+          hash
         );
 
+        const response = await blobClient.download(0);
+
         const blobReadableStream = response.readableStreamBody;
+        if (!blobReadableStream) {
+          throw new Error("Unable to fetch blob.");
+        }
+
+        fs.mkdirpSync(temporaryBlobOutputFolder);
+
         const tarWritableStream = tar.extract({
-          cwd: parentFolderWhereToExtractFolder
+          cwd: temporaryBlobOutputFolder
         });
 
-        if (!blobReadableStream) {
-          throw new Error("Unable to fetch blob.");
+        blobReadableStream.pipe(tarWritableStream);
+
+        const blobPromise = new Promise((resolve, reject) => {
+          blobReadableStream.on("end", () => resolve());
+          blobReadableStream.on("error", error => reject(error));
+        });
+
+        await blobPromise;
+      } catch (error) {
+        fs.removeSync(temporaryBlobOutputFolder);
+
+        if (error && error.statusCode === 404) {
+          return;
+        } else {
+          throw new Error(error);
         }
+      }
+    }
 
-        return blobReadableStream.pipe(tarWritableStream);
-      })
-      .then(() => true)
-      .catch(() => false);
+    return temporaryBlobOutputFolder;
```

Comment:
if we get back a 404 it still returns the temporaryBlobOutputFolder path which refers to a folder that does not exist, it this correct?

Reply:
No. If we get a 404 back it returns undefined (line 81) which means the parent will return false.